rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,234 @@
+This makes it much easier to contribute new packages packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `pkgs/top-level/all-packages.nix`.
+
+# Motivation
+[motivation]: #motivation
+
+- (Especially new) package contributors are having a hard time figuring out which files to add or edit. These are very common questions:
+  - Which directory should my package definition go in?
+  - What are all the categories and do they matter?
+  - What if the package has multiple matching categories?
+  - Why can't I build my package after adding the package file?
+  - Where in all-packages.nix should my package go?
+- Figuring out where an attribute is defined is a bit tricky:
+  - First one has to find the definition of it in all-packages.nix to see what file it refers to
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even deadlocks editors due to the file size
+  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
+- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC establishes the standard of using `pkgs/unit/${shard}/${name}` "unit" directories for the definitions of the Nix packages `pkgs.${name}` in nixpkgs, where `shard = toLower (substring 0 2 name)`.
+All unit directories are automatically discovered and incorporated into the `pkgs` set using `pkgs.${name} = pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix { }`.
+
+The following requirements will be checked by CI.
+This standard must be followed for newly added packages that can satisfy these requirements.
+A treewide migration to this standard will be performed for existing packages that can satisfy these requirements.
+
+## Structure
+
+The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${shard}/${name}`.
+Each unit directory must contain at least a `pkg-fun.nix` file, but may contain arbitrary other files and directories.
+
+This ensures that maintainers don't have to verify this structure manually, which is prone to mistakes.
+
+## Only derivations
+
+If `pkgs/unit/${shard}/${name}` exists, `pkgs.${name}` must be a derivation that can be built directly with `nix-build`.
+
+This ensures that people can expect the unit directories to correspond to buildable packages and not functions like `pkgs.fetchFromGitHub` or `pkgs.buildRustCrate`.
+
+## Stable boundary
```

Comment:
How do you plan to enforce this? No ofborg check, no boundary.

Reply:
Tooling to enforce this is being worked on [here](https://github.com/nixpkgs-architecture/nix-spp).

Roughly it works by parsing all Nix files in nixpkgs, figuring out which other files they (statically) reference, and erroring when those references cross unit boundaries. This will be the same code that performed the auto-migration in [the draft implementation](https://github.com/NixOS/nixpkgs/pull/211832), which notably skipped packages that couldn't be safely moved without breaking references.

It's not 100% perfect because files can also be referenced dynamically, e.g. this (made-up) function `subpath: import (dirOf ./. + "subpath")`, but within nixpkgs static references (like `import ./some/file.nix`) are most common, so it works out.